File: tests/Parity/compiler-support/cases.py
#!/usr/bin/env python3
import json, os, sys

SYNTHETIC = [
    '', '<script src="/app.js" defer></script>',
    '<style scoped>.x { color:red }</style><link REL="preload stylesheet" href="/a.css" media="screen">',
    '<script>import x from "x";\nconst a = 1;\nexport default { mounted() { return a }, value: 2 }</script>',
    '<script setup lang="ts">const x: number = 1; export default { async save() {}, run: (x) => x }</script>',
    '<script src="{{ asset(\'x.js\') }}" id="main" class="foo" async></script>',
    '<style id="theme" nonce="abc">body { color: blue }</style>',
]

for i, source in enumerate(SYNTHETIC, 1):
    print(json.dumps({'fn':'register','args':[source], 'name':f'synthetic-{i:02d}'}, ensure_ascii=False))

def root():
    p=os.path.abspath(__file__)
    while p != os.path.dirname(p):
        p=os.path.dirname(p)
        if os.path.isdir(os.path.join(p,'builder','.reference','python','src')): return p
    raise RuntimeError

repo=root(); count=0
for base in [os.path.join(repo,'saola','resources'), os.path.join(repo,'php-compiler','tests','Parity','source-split','fixtures')]:
    if not os.path.isdir(base): continue
    for directory, _, names in os.walk(base):
        for name in sorted(names):
            if not name.endswith('.sao'): continue
            path=os.path.join(directory,name)
            source=open(path,encoding='utf-8-sig').read()
            print(json.dumps({'fn':'register','args':[source], 'name':os.path.relpath(path,repo)}, ensure_ascii=False))
            count += 1

utility_calls = [
    ('format_fetch_config', [None]), ('format_fetch_config', [{}]),
    ('format_fetch_config', [{'url':'`/users/${id}`','method':'POST','headers':{'X-Test':'✓'},'cache':False,'limit':2}]),
    ('format_attrs', [{'a':'x','enabled':True,'n':2}]),
    ('format_attributes_to_json', [None]),
    ('format_attributes_to_json', [{'defer':True,'nonce':123,'name':'x'}]),
]
for i,(fn,args) in enumerate(utility_calls,1):
    print(json.dumps({'fn':fn,'args':args,'name':f'utils-{i:02d}'},ensure_ascii=False))
print(f'  7 ca tổng hợp + {count} source thật + {len(utility_calls)} utility', file=sys.stderr)

File: tests/Parity/main-compiler/cases.js
#!/usr/bin/env node
'use strict';
const fs=require('fs'),path=require('path');
function root(){let d=__dirname;while(d!==path.dirname(d)){d=path.dirname(d);if(fs.existsSync(path.join(d,'builder','src','index.js')))return d;}throw new Error('root');}
const repo=root(),Compiler=require(path.join(repo,'builder','src','index.js')),compiler=new Compiler();
const synthetic=[
 ['plain','<template>\n<div>Hello</div>\n</template>'],
 ['props',"@props({ title: 'Hello', count: 2 })\n<template>\n<h1>{{ title }}</h1>\n</template>"],
 ['states',"@states({ count: 0 })\n@computed(double = $count * 2)\n<template>\n<button @click($count++)>{{ $double }}</button>\n</template>"],
 ['extends',"@extends('layouts.app', ['x' => 1])\n<template>\n@block('body')\n<p>Body</p>\n@endblock\n</template>"],
 ['typescript','<script setup lang="ts">\ninterface Item { id: number }\nconst local: Item = { id: 1 };\nexport default { mounted() {}, save: (x: number) => x }\n</script>\n<template>\n<div>{{ local.id }}</div>\n</template>'],
 ['assets','<script src="{{ asset(\'app.js\') }}" defer></script>\n<style scoped>.x { color:red }</style>\n<link rel="stylesheet" href="/app.css">\n<template><div class="x">x</div></template>'],
 ['loops',"@props({ items: [] })\n<template>\n@foreach(items as item)\n<button @click.prevent(select(item.id))>{{ item.name }}</button>\n@endforeach\n</template>"],
 ['wrapper',"<template>\n@wrapper('section', ['class' => 'page'])\n<div>inside</div>\n@endwrapper\n</template>"],
 ['verbatim',"<template>\n@verbatim\n<code>`x ${y}` and 'quote'</code>\n@endverbatim\n</template>"],
 ['await','@props({ users: [] })\n@await\n@fetch(\'/users\')\n<template>\n<div>{{ users.length }}</div>\n</template>'],
];
let index=0,real=0;
function pascal(s){return s.split(/[-_\s]+/).map(w=>w.charAt(0).toUpperCase()+w.slice(1)).join('');}
function emit(name,source,file){
 const parts=compiler.parseSaoFile(source,file||name+'.sao');
 const blade=compiler.preprocessor.preprocess(parts,{assetPrefix:'static/parity/assets/'});
 const scripts=parts.cleanedContent.match(/<script[^>]*>[\s\S]*?<\/script>/gi)||[];
 const styles=parts.cleanedContent.match(/<style[^>]*>[\s\S]*?<\/style>/gi)||[];
 const links=parts.cleanedContent.match(/<link\b(?=[^>]*\brel\s*=\s*["'][^"']*\bstylesheet\b[^"']*["'])[^>]*>/gi)||[];
 let code=blade.declarations.length?blade.declarations.join('\n')+'\n\n':'';
 if(scripts.length||styles.length||links.length)code+=[...scripts,...styles,...links].join('\n')+'\n\n';
 code+=parts.wrapperType?`<${parts.wrapperType}>\n${blade.blade}\n</${parts.wrapperType}>`:blade.blade;
 const component='Case'+String(++index).padStart(3,'0'),view='parity.'+pascal(name.replace(/[^A-Za-z0-9_-]+/g,'-'));
 process.stdout.write(JSON.stringify({name,code,view,functionName:component,factoryName:'Parity'+component})+'\n');
}
for(const [name,source] of synthetic)emit(name,source);
function walk(dir){if(!fs.existsSync(dir))return;for(const e of fs.readdirSync(dir,{withFileTypes:true}).sort((a,b)=>a.name.localeCompare(b.name))){const f=path.join(dir,e.name);if(e.isDirectory())walk(f);else if(e.name.endsWith('.sao')&&e.name!=='13-unclosed.sao')try{emit(path.relative(repo,f),fs.readFileSync(f,'utf8'),f);real++;}catch(_){}}}
walk(path.join(repo,'saola','resources'));walk(path.join(repo,'php-compiler','tests','Parity','source-split','fixtures'));
process.stderr.write(`  ${synthetic.length} ca tổng hợp + ${real} source thật\n`);
